src/test/regress/sql/multi_hash_pruning.sql

```diff
@@ -284,5 +304,29 @@ SELECT count(*) FROM orders_hash_partitioned
 SELECT count(*) FROM orders_hash_partitioned
 	WHERE o_orderkey IN (1,2) OR o_custkey NOT IN (SELECT o_custkey FROM orders_hash_partitioned WHERE o_orderkey = 3);
 
+-- left joins should prune shards based on the left hand side of the left join
+-- it should only assign 2 tasks as there is a filter on the left table pruning to 2
+-- shards
+SELECT count(*)
+FROM orders_hash_partitioned
+         LEFT JOIN lineitem_hash_partitioned ON (o_orderkey = l_orderkey)
+WHERE o_orderkey IN (1, 2);
+
+-- same principle but on a right join
+SELECT count(*)
+FROM orders_hash_partitioned
+RIGHT JOIN lineitem_hash_partitioned ON (o_orderkey = l_orderkey)
+WHERE l_orderkey IN (1, 2);
+
+-- full outerjoin should not prune based on filters of either side as missing shards
+-- create empty rows on the other side
+SELECT count(*)
+FROM orders_hash_partitioned
+FULL OUTER JOIN lineitem_hash_partitioned ON (o_orderkey = l_orderkey)
+WHERE o_orderkey IN (1, 2)
+   OR l_partkey = 7; -- causes lineitems to yield tuples irrespective of orders, would have been optimized to a leftjoin without
```

Comment:
This might actually be the one I was looking for on Friday, will add

Reply:
I have added tests, but due to how postgres pushes down filters it will not prune anything with the `OR` filter. Also added a test for an `AND` filter which I believe should be possible to prune to 1 shards, but postgres instead prunes to 3 shards.

For shard pruning we heavily rely on postgres pushing down filters into the table access. To get these pruned to the minimal viable shards might require an overhaul of how shards get pruned to have a better understanding of the join clause in combination with AND/OR clauses.